<center> <img src="https://upload.wikimedia.org/wikipedia/commons/c/c8/Iteso_logo.jpg" align="center" width="200" height="200"/> </center>

<center><font color=#555555><font size=6> Second Laboratory </font> <br> <br>
    
<center><font color=#555555><font size=5> Microstructures and Trading Systems </font> <br> <br> <br> <br>

<center><font color=#555555><font size=6> Bryan Manica Pineda</font> <br> <br>
    
<center><font color=#555555><font size=4> if722176@iteso.mx</font> <br> <br>
   
<center> <font color= #555555> <font size = 4> June. 2022 </a> | <font color= #555555> <font size = 4> Repository: <a href='https://github.com/bmanica/aptmodel-lab2'>Link</a></font>

<hr style="border:0.02in solid gray"> </hr>

<center> <font color= #555555> <font size = 6> High Frequency Models Testing </font> <br> <br> <font color= #555555> <font size = 5> APT and Roll Model into real world data  </font>

<hr style="border:0.02in solid gray"> </hr>

<font color= #555555> <font size = 4> Abstract </font>

<font color= #6B6B6B> <font size = 3>  Include here a description of the whole work, use no more than 400 words, you can use **Bold** and *Italic* highlights, also you can inclue [links](www.iteso.mx). Usually is a good strategy to generate this text at the very end of the process, since you have to include all aspects of the work, from a brief introduction, to problem definitions and methods used, to finally the results found. </font>

<br>

# <font color= #6B6B6B> <font size = 6> 1. Introduction </font>

<hr style="border:0.01in solid gray"> </hr>

This is an introduction, normally, it is done at the very end, when all the other content in the notebook is completed but just before the abstract (which is short version of the introduction) ... 

<br>

# <font color= #6B6B6B> 2. Install/Load Packages and Depedencies </font>

<hr style="border:0.01in solid gray"> </hr>

## <font color= #6B6B6B> 2.1 Python Packages </font> <font color= #555555>

Para poder correr el presente **notebook** es necesario haber instalado todas aquellas librerias que se encuentran descritas en el archivo adjunto en el repositorio con nombre _requirements.txt_. Dichas dependencias son las siguientes:

- **pandas** >= 1.1.0
- **numpy** >= 1.19.1
- **jupyter** >= 1.0.0
- **chart_studio** >= 1.1
- **plotly** >= 4.14

## <font color= #6B6B6B> 2.2 Files Dependencies </font> <font color= #555555>

Además de todas las librerias previamente mencionadas, el correcto funcionamiento de este **notebook** también requiere de algunos datos externos. Estos archivos se encuentran directamente en la carpeta _files_ adjunta dentro de este mismo repositorio:

- files/orderbooks_05jul21.json: Archivo tipo json. Datos históricos de diversos orderbooks correspondientes a una hora el 5 de julio del 2021.

## <font color= #6E6E6E> 2.3 Install Packages</font>

En caso de que no se hayan instalado los paquetes dentro del ambiente virtual sobre el cual se está trabajando, se pueden instalar directamente dentro de este cuaderno. A continuación se encuentra la estructura de comandos necesarias para instalar las librerias, solo correr en caso de que no se hayan instalado previamente:

In [1]:
%%capture

# Install all the pip packages in the requirements.txt
import sys
!{sys.executable} -m pip install -r requirements.txt

## <font color= #6E6E6E> 2.3 Load Packages and Scripts</font>

Ya con los requerimentos computacionales bien definidos, ahora se proseguirá a cargar tanto las librerias como los scripts de python 3 para poder realizar el análisis correspondiente.

In [1]:
### Required packages
import pandas as pd
import numpy as np

### Required local scripts
import data as dt
import functions as fn
import visualizations as vz

<br>

# <font color= #6B6B6B> 3. Data Description </font>

<hr style="border:0.01in solid gray"> </hr>

Para poder realizar este laboratorio en particular, se nos fueron proporcionados los datos previamente mencionados. A continuación se comenzará a describir un poco más a detalle la estructura de los mismos para entender de mejor manera la materia prima sobre la cual se estará trabajando.

**Order Book.**

Este conjunto de datos se encuentra bajo el formato _json_, por tanto se compone por una lista de diccionarios donde cada subconjuto de valores responde a un libro de órdenes del activo con clave de pizarra **_BTCUSDT_**, dicho orderbook se encuentra asociado a cada punto del tiempo que representa la llave del diccionario. La actualización de dichos libros se da en términos de segundos, por lo tanto la ventana de tiempo sobre la cual trabajaremos es de un poco más de $1$ hora. A pesar de que pareciera que no tenemos tantos datos, la realidad es que cada libro cuenta con una cantidad suficiente de datos para poder extraer información relevante a ese punto del tiempo que se encuentra asociado.

## <font color= #6B6B6B> 3.1. Order Book </font>

Contamos entonces con una lista de diccionarios donde cada llave representa un punto en el tiempo donde se contaba con ese determinado libro de órdenes, ahora lo que haremos será ir definiendo y mostrando un poco mas sobre los datos ya desde una perspectiva un poco más visual.

In [2]:
### Let's visualize the keys range.
ob_data = dt.ob_data_bit
range_ob = pd.to_datetime(pd.Series(list(ob_data.keys()))).sort_values()
print(f'El primer tiempo registrado está en: {range_ob[0]}, y el último en: {range_ob.tail(1).item()}')

El primer tiempo registrado está en: 2021-07-05 13:06:46.571000+00:00, y el último en: 2021-07-05 14:06:46.412000+00:00


Como podemos observar ya directamente de la descripción de los rangos para las llaves de los orderbooks contamos con un poco menos de $1$ hora de información correspondiente al día **2021-07-05** entre $1$ y $2$ de la tarde.

In [3]:
### Let's see how many books do we have
print(f'En una hora de tiempo contamos con {len(list(ob_data.keys())):,} libros')

En una hora de tiempo contamos con 2,401 libros


Entonces si bien pareciera que $1$ hora de tiempo no puede aportar mucha información, la realidad es que dentro de este horario contamos con alrededor de $2,401$ libros. A continuación se mostrará un poco respecto al contenido de cada orderbook, esto con la intención de alcanzar a percibir la cantidad de datos que tenemos.

In [4]:
### Let's see the first orderbook values
print(f'Este primer orderbook corresponde al punto en el tiempo. {range_ob[0]}:')
list(ob_data.values())[0].head(5)

Este primer orderbook corresponde al punto en el tiempo. 2021-07-05 13:06:46.571000+00:00:


bid_size      bid      ask  ask_size
0  0.000400  28270.0  28275.0  0.025405
1  0.009787  28269.0  28276.0  0.516810
2  0.008168  28268.0  28277.0  0.005044
3  0.995787  28266.0  28278.0  0.377374
4  1.038704  28265.0  28280.0  1.179715

Como podemos percibir del dataframe correspondiente al primer punto del tiempo, contamos con diferentes atributos:

- **bid_size**: Corresponde al volumen posicionado al del **bid** asociado.
- **bid**: Responde directamente al precio más alto que un comprador esta dispuesto a pagar por el activo de interés.
- **ask**: Es el precio más bajo al cual un vendedor esta dispuesto a vender el activo.
- **ask_size**: Corresponde al volumen posicionado al del **ask** asociado.

Todas y cada una de las llaves correspondientes a los datos del orderbook contienen un dataframe con las mismas características que se encuentra desplegado. Entonces a partir de estos valores es que se estará experimentando y observando la eficacia de los modelos de alta frecuencia propuestos.

<br>

# <font color= #6B6B6B> 4. APT Model Testing </font>

<hr style="border:0.01in solid gray"> </hr>

Con la intención de poder eficientar tiempos computacionales se decide trabajar sobre una serie de _scripts_ de **python 3**, los cuales se encuentran anexos al repositorio <font color= #555555> <a href='https://github.com/bmanica/aptmodel-lab2'>aptmodel-lab2</a></font>. En ellos se encuentra predefinida una serie de cálculos los cuales nos ayudarán a realizar el test y analizar el cumplimiento del modelo frente a situaciones de la vida real.

Antes de poder entrar en detalle respecto a la serie de pruebas realizadas para el **APT**, es importante tener claro el origen y la teoría dentrás de este modelo de alta frecuencia. En este sentido el modelo parte de su definición básica _"poder encontrar el pago futuro respecto a un activo financiero que presenta flujos"_, de manera que la ecuación básica para partir es la siguiente:

$$x_{t+1}=p_{t+1}+d_{t+1}$$

donde:

- $x_{t+1}: \text{Pago del activo a realizarse a un momento futuro}$
- $p_{t+1}: \text{Precio futuro de la acción}$
- $d_{t+1}: \text{Dividendo que se espera pague la acción a un momento futuro}$

Ahora hasta el momento se encuentra definido el pago que se podría esperar en un momento $t+1$, sin embargo el concepto económico que nos ayuda  

In [5]:
### Let's define the apt testing data frames.
apt_all = fn.apt_check_all(ob_data)
apt_tob = fn.apt_check_tob(ob_data)

print(f'The keys for apt model test with all orders are: {list(apt_all.keys())}')
print(f'The keys for apt model test with tob orders are: {list(apt_tob.keys())}')

The keys for apt model test with all orders are: ['simple_mid_price', 'weighted_mid_price']
The keys for apt model test with tob orders are: ['simple_mid_price', 'weighted_mid_price_a', 'weighted_mid_price_b']


## <font color= #6B6B6B> 4.1. Testing All Orders </font>

Este primer enfoque que se propone analizar incluye todas las órdenes de todos los libros de órdenes presentes en el diccionario de entrada, es decir, para este primer punto de análisis queremos observar los cambios en los precios de prácticamente todas las potenciales operaciones de los orderbooks (aproximadamente unos $60,000$ registros a consolidar). La metodología que se propone es la misma que se menciona en la definición del modelo, se busca encontrar la cantidad de veces que el **Mid-Price** sigue un proceso estocástico tipo _Martingala_ para cada minuto en el tiempo registrado. 

### <font color= #6B6B6B> 4.1.1. Simple Mid-Price </font>

Para este análisis se calcula es **Simple Mid-Price** el cual se define matemáticamente por la siguiente expresión:

$$\text{Mid-Price} = \frac{\text{bid} + \text{ask}}{2}$$

Y es este valor el que se calcula para todas y cada una de las órdenes presentes en cada orderbook del rango de datos con el que se cuenta.

In [23]:
### Let's see the data frame for simple mid-price test
print(f'''The min proportion where the mid-price follows martingale process is {apt_all['simple_mid_price']['P. Exp 1'].min()}
and it happened on {apt_all['simple_mid_price']['P. Exp 1'].idxmin()}''')

apt_all['simple_mid_price'].head()

The min proportion where the mid-price follows martingale process is 0.67
and it happened on 2021-07-05 13:31:00


Exp 1  Exp 2  P. Exp 1  P. Exp 2
Time                                                 
2021-07-05 13:06:00    177     47      0.79      0.21
2021-07-05 13:07:00    781    243      0.76      0.24
2021-07-05 13:08:00    799    200      0.80      0.20
2021-07-05 13:09:00    750    224      0.77      0.23
2021-07-05 13:10:00    789    235      0.77      0.23

### <font color= #6B6B6B> 4.1.2. Weighted Mid-Price </font>

Para términos del **Weighted Mid-Price** se calculó a través de la siguiente expresión matemática:

$$\text{Weighted Mid-Price} = \bigg(\frac{\text{bv}}{\sum_{i=1}^{n}\big(\text{bv}_{i}+\text{av}_{i}\big)} \bigg) \text{ap} + \bigg(\frac{\text{av}}{\sum_{i=1}^{n}\big(\text{bv}_{i}+\text{av}_{i}\big)} \bigg) \text{bp}$$

donde:

- $\text{bv}: \text{bid volume}$
- $\text{av}: \text{ask volume}$
- $\text{ap}: \text{ask price}$
- $\text{bp}: \text{bid price}$

En este sentido y al estar tomando practicamente todos las órdenes presentes, la profundidad detonada por $i$ solo se define para su mismo nivel.

In [24]:
### Let's see the data frame for weighted mid-price test
print(f'''The max proportion where the mid-price follows martingale process is {apt_all['weighted_mid_price']['P. Exp 1'].max()}
and it happened on {apt_all['weighted_mid_price']['P. Exp 1'].idxmax()}''')

apt_all['weighted_mid_price'].head()

The max proportion where the mid-price follows martingale process is 0.01
and it happened on 2021-07-05 14:02:00


Exp 1  Exp 2  P. Exp 1  P. Exp 2
Time                                                 
2021-07-05 13:06:00      0    224       0.0       1.0
2021-07-05 13:07:00      0   1024       0.0       1.0
2021-07-05 13:08:00      0    999       0.0       1.0
2021-07-05 13:09:00      0    974       0.0       1.0
2021-07-05 13:10:00      0   1024       0.0       1.0

## <font color= #6B6B6B> 4.2. Testing Top of the Book Orders </font>

En este segundo enfoque de análisis se busca realizar un proceso análogo al previamente descrito, sin embargo, la principal diferencia para este punto se da en que ahora solo se utilizarán los **Mid-Prices** que se definen única y exclusivamente para el _Top of the Book_, es decir, esta nueva propuesta solo tomará un precio para cada libro de órdenes, de manera que el procesamiento de datos se reducirá considerablemente ya que ahora solo se contarán con $2,400$ registros a consolidar por minuto.

### <font color= #6B6B6B> 4.2.1. Simple Mid-Price </font>

Para el cálculo de este **Simple Mid-Price** se define prácticamente la misma fórmula que se utiliza para todas las órdenes, la diferencia principal radica en que ahora solo se tomará el _Top of the Book_, en otras palabras este nuevo análisis solo tomará el mejor **bid** y el mejor **ask** presente en cada libro de órdenes.

En este sentido y tratando de llegar a un análisis comparativo entre los **Mid-Prices** del _Top of the Book_ y el resto de las órdenes, podríamos comenzar a intuir que las proporciones donde se cumpla que el precio sigue una martingala con esta nueva estructura de análisis será menor, ya que al tomar siempre el _Top of the Book_ podríamos decir que estaríamos más cerca de "definir" la tendencia del precio a cada punto del tiempo, y en ese sentido se podría esperar que las fluctuaciones en el precio fueran mayores en comparación a si tomamos todas las órdenes presentes.

In [25]:
### Now let's render the simple mid-price experiment just for tob
print(f'''The min proportion where the mid-price follows martingale process is {apt_tob['simple_mid_price']['P. Exp 1'].min()}
and it happened on {apt_tob['simple_mid_price']['P. Exp 1'].idxmin()}''')

apt_tob['simple_mid_price'].head()

The min proportion where the mid-price follows martingale process is 0.65
and it happened on 2021-07-05 13:52:00


Exp 1  Exp 2  P. Exp 1  P. Exp 2
Time                                                 
2021-07-05 13:06:00      6      3      0.67      0.33
2021-07-05 13:07:00     27     14      0.66      0.34
2021-07-05 13:08:00     31      9      0.78      0.22
2021-07-05 13:09:00     27     12      0.69      0.31
2021-07-05 13:10:00     30     11      0.73      0.27

In [55]:
### Let's see the mean proportion for experiment 1.
print(f'''With all orders we have a mean proportion of {round(apt_all['simple_mid_price']['P. Exp 1'].mean(),4)*100}%''')
print(f'''With tob orders we have a mean proportion of {round(apt_tob['simple_mid_price']['P. Exp 1'].mean(),4)*100}%''')

With all orders we have a mean proportion of 80.41%
With tob orders we have a mean proportion of 72.39%


### <font color= #6B6B6B> 4.2.2. Weighted Mid-Price </font>

In [26]:
### Let's see now the tob experiments for the first way of calculation of weighted mid-price
print(f'''The min proportion where mid-price follows martingale process is {apt_tob['weighted_mid_price_a']['P. Exp 1'].min()}
and it happened on {apt_tob['weighted_mid_price_a']['P. Exp 1'].idxmin()}''')

apt_tob['weighted_mid_price_a'].head()

The min proportion where mid-price follows martingale process is 0.65
and it happened on 2021-07-05 13:08:00


Exp 1  Exp 2  P. Exp 1  P. Exp 2
Time                                                 
2021-07-05 13:06:00      6      3      0.67      0.33
2021-07-05 13:07:00     27     14      0.66      0.34
2021-07-05 13:08:00     26     14      0.65      0.35
2021-07-05 13:09:00     26     13      0.67      0.33
2021-07-05 13:10:00     27     14      0.66      0.34

### <font color= #6B6B6B> 4.2.3. Comparing with alternative Weighted Mid-Price </font>

In [27]:
### Now let's compare it with the alternative way of calculating weighted mid-price
print(f'''The min proportion where mid-price follows martingale process is {apt_tob['weighted_mid_price_b']['P. Exp 1'].min()}
and it happened on {apt_tob['weighted_mid_price_b']['P. Exp 1'].idxmin()}''')

apt_tob['weighted_mid_price_b'].head()

The min proportion where mid-price follows martingale process is 0.65
and it happened on 2021-07-05 13:08:00


Exp 1  Exp 2  P. Exp 1  P. Exp 2
Time                                                 
2021-07-05 13:06:00      6      3      0.67      0.33
2021-07-05 13:07:00     27     14      0.66      0.34
2021-07-05 13:08:00     26     14      0.65      0.35
2021-07-05 13:09:00     26     13      0.67      0.33
2021-07-05 13:10:00     27     14      0.66      0.34

In [56]:
### Let's see the cases where the weighted mid prices doesnt match
sum(apt_tob['weighted_mid_price_a']['P. Exp 1'] != apt_tob['weighted_mid_price_b']['P. Exp 1'])

25

<br>

# <font color= #6B6B6B> 5. Roll Model Testing  </font>

<hr style="border:0.01in solid gray"> </hr>

<br>

# <font color= #6B6B6B> 5. Results and Visual Resources  </font>

<hr style="border:0.01in solid gray"> </hr>

## <font color= #6B6B6B> 5.1 Results of first processes </font>

In [6]:
# DELETE THIS CELL IN YOUR FILE

# Separate your results according to the main processes you are working on

<br>

## <font color= #6B6B6B> 5.2 Results of another proces </font>

In [7]:
# DELETE THIS CELL IN YOUR FILE

# You can have several processes with which you are obtaining the results you need.

<br>

# <font color= #6B6B6B> 6. Conclusions  </font>

<hr style="border:0.01in solid gray"> </hr>

<br>

# <font color= #6B6B6B> 7. References </font>

<hr style="border:0.01in solid gray"> </hr>

[1] Munnoz, 2020. Python project template. https://github.com/iffranciscome/python-project. (2021).

<br>